lib/SPIRV/SPIRVWriter.cpp

```diff
@@ -2147,8 +2164,22 @@ void LLVMToSPIRV::transFunction(Function *I) {
     }
   }
 
-  if (BF->getModule()->isEntryPoint(spv::ExecutionModelKernel, BF->getId()) &&
-      BF->shouldFPContractBeDisabled()) {
+  bool IsKernelEntryPoint =
+      BF->getModule()->isEntryPoint(spv::ExecutionModelKernel, BF->getId());
+  bool DisableContraction = false;
+  switch (FPContractMode) {
+  case FPContractFast:
+    DisableContraction = false;
+    break;
+  case FPContractOn:
+    DisableContraction = IsKernelEntryPoint && BF->shouldFPContractBeDisabled();
+    break;
+  case FPContractOff:
+    DisableContraction = IsKernelEntryPoint;
+    break;
+  }
+
+  if (DisableContraction) {
     BF->addExecutionMode(BF->getModule()->add(
         new SPIRVExecutionMode(BF, spv::ExecutionModeContractionOff)));
   }
```

Comment:
Github doesn't allow me to left a comment to completely unchanged code, but could you please modify line 2186 so it re-uses your `IsKernelEntryPoint` variable?

Reply:
Done.